qark/decompiler/decompiler.py

```diff
@@ -0,0 +1,228 @@
+import logging
+from multiprocessing.pool import ThreadPool
+import os
+import platform
+import shlex
+import shutil
+import subprocess
+import re
+import zipfile
+
+log = logging.getLogger(__name__)
+
+JAVA_VERSION_REGEX = '"(\d+\.\d+\..+)\"'
+LIB_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "../../lib")
+APK_TOOL_PATH = os.path.join(LIB_PATH, "apktool")
+DEX2JAR_PATH = os.path.join(LIB_PATH, "dex2jar-2.0")
+DECOMPILERS_PATH = os.path.join(LIB_PATH, "decompilers")
+
+APK_TOOL_COMMAND = ("java -Djava.awt.headless=true -jar {apktool_path}/apktool.jar "
+                    "d {path_to_apk} --no-src --force -m --output {build_directory}")
+DEX2JAR_COMMAND = "{dex2jar_path}/d2j-dex2jar.sh {path_to_dex} -o {build_directory}/{apk_name}.jar"
+DECOMPILER_COMMANDS = {"JDCORE": "java -jar {path_to_decompiler} {jar} {build_directory}/jdcore",
+                       "PROCYON": "java -jar {path_to_decompiler} {jar} -o {build_directory}/procyon",
+                       "CFR": "java -jar {path_to_decompiler} {jar} --outputdir {build_directory}/cfr"}
+
+
+class Decompiler(object):
+    """This class handles unpacking and decompiling APKs into Java source code.
+
+    New compilers can be added by adding the command to run the compiler to the `DECOMPILER_COMMANDS` dictionary and
+    adding them into the `Decompiler.decompilers` dictionary with the path to the decompiler on the system.
```

Comment:
> +    New compilers can be added by adding the command to run the compiler to the `DECOMPILER_COMMANDS` dictionary and
> +    adding them into the `Decompiler.decompilers` dictionary with the path to the decompiler on the system.

These seems inextricably linked and should just go in a small helper class. Something along the lines of:
```
class ExternalDecompiler(object):
    def __init__(self, name, local_path, command=None):
        self.name = name.lower()
        self.local_path = local_path
        self.command = command or 'java -jar {local_path} {build_directory}/{name}'.format(local_path=self.local_path, build_directory='{build_directory}', name=self.name)

jd_core = ExternalDecompiler('JDCORE', '/path/to/jdcore')

list_of_decompilers = [jd_core]

decompiler = Decompiler(decompilers=list_of_decompilers)
```

Or you could make ExternalDecompiler an ABC and subclass it for each decompiler. Either way, I think a helper class is the way to go.

Reply:
This is fixed in the newest commit. ExternalDecompiler is made into an ABC.